### PR TITLE
Add unsubscribe url endpoint

### PR DESCRIPTION
**Pull request by BlessedDev**
This PR based builds the one-click unsubscribe endpoint to unsubscribe end users as described in [this](https://trello.com/c/qFVyAmMk/32-build-notificationsservicegovuk-unsubscribe) card.
It stops short of "batching" unsubscribe requests as that is a separate piece of work.

### CODE REVIEW

File: tests/app/one-click-unsubscribe/test_one_click_unsubscribe.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    token = generate_token(
+        sample_email_notification.to, current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"]
+    )
+    response = unsubscribe_url_post(client, sample_email_notification.id, token)
+    response_json_data = response.get_json()
+    created_unsubscribe_request = get_unsubscribe_request_by_notification_id_dao(sample_email_notification.id)
+    assert response.status_code == 200
+    assert response_json_data["message"] == "Unsubscribe successful"
+    assert response_json_data["result"] == "success"
+    assert created_unsubscribe_request.notification_id == sample_email_notification.id
```
Comment (leohemsted) [inline tests/app/one-click-unsubscribe/test_one_click_unsubscribe.py:23]: i think id like a couple of more thorough assertions here on the other values of the unsubscribe request

File: app/one_click_unsubscribe/rest.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    notification = get_notification_by_id(notification_id)
+    create_unsubscribe_request_dao(notification)
+
+    current_app.logger.debug("Received unsubscribe request for notification_id: %s", notification_id)
+
+    return jsonify(result="success", message="Unsubscribe successful"), 200
+
+
+def generate_unsubscribe_url_token(data):
+    generate_token(str(data), current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"])
```
Comment (leohemsted) [inline app/one_click_unsubscribe/rest.py:35]: this should return - probably need some more test if we didnt catch this yet

**Review by leohemsted — COMMENTED** [review]
a few comments - i also wonder if we need a bit more test coverage here, mostly around making sure the unsubscribe request is stored in the db correctly, and the relationships work. eg: does it still work if theres a more recent template version, or the template is archived, etc? Are the created_at etc timestamps correct?

File: app/one_click_unsubscribe/rest.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    # Create unsubscribe request from the notification object
+    notification = get_notification_by_id(notification_id)
+    create_unsubscribe_request_dao(notification)
+
+    current_app.logger.debug("Received unsubscribe request for notification_id: %s", notification_id)
+
+    return jsonify(result="success", message="Unsubscribe successful"), 200
+
+
+def generate_unsubscribe_url_token(data):
```
[the remark this answers is not in the document]
Comment (BlessedDev) [reply app/one_click_unsubscribe/rest.py:34]: This is an error, It was part of a half thought process I had abandoned. The tests didn't catch it as it is not being used anywhere. I had intended to take it out and forgot.
I will add a `generate_unsubscribe_link_url` function as part of the work in [this](https://trello.com/c/6H3orCLk/33-add-list-unsubscribe-headers-to-emails-if-the-template-is-unsubscribable) ticket.

File: tests/app/one-click-unsubscribe/test_one_click_unsubscribe.py (hunk 1 above)
Comment (BlessedDev) [reply tests/app/one-click-unsubscribe/test_one_click_unsubscribe.py:23]: Good point. I will add a few more including ` template_version`

File: app/one_click_unsubscribe/rest.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    try:
+        check_token(token, current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"], max_age_seconds)
+
+    except BadData as e:
+        errors = {"unsubscribe request": "This is not a valid unsubscribe link."}
+        raise InvalidRequest(errors, status_code=400) from e
+
+    # Create unsubscribe request from the notification object
+    notification = get_notification_by_id(notification_id)
```
[the remark this answers is not in the document]
Comment (BlessedDev) [reply app/one_click_unsubscribe/rest.py:26]: Good catch, I will deal with that scenario.

**Comment by BlessedDev** [comment]
> a few comments - i also wonder if we need a bit more test coverage here, mostly around making sure the unsubscribe request is stored in the db correctly, and the relationships work. eg: does it still work if theres a more recent template version, or the template is archived, etc? Are the created_at etc timestamps correct?

I have added tests for two test cases:
i. the template is updated
ii. the template is redacted

I have made some assumptions in the tests and hopefully they are accurate.

File: tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    notification = create_notification(template=test_template, to_field="foo@bar.com")
+    initial_template_version = test_template.version
+    token = generate_token(notification.to, current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"])
+
+    response = unsubscribe_url_post(client, notification.id, token)
+
+    # update template content to generate new template version
+    test_template.content = "New content"
+    test_template.process_type = "priority"
+    dao_update_template(test_template)
```
Comment (leohemsted) [inline tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py:47]: should this step should happen before we call `unsubscribe_url_post`?

so in order:

* create notification
* update template
* unsubscribe endpoint is called

File: tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    initial_template_version = test_template.version
+    notification = create_notification(template=test_template, to_field="foo@bar.com")
+    token = generate_token(notification.to, current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"])
+
+    response = unsubscribe_url_post(client, notification.id, token)
+
+    # archive template
+    test_template.archived = True
+    dao_update_template(test_template)
+    subsequent_template_version = test_template.version
```
Comment (leohemsted) [inline tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py:74]: same point about this happening out of order as above

**Review by leohemsted — COMMENTED** [review]
looks pretty good! just a couple of minor points about the order of some events in the tests and then i think this is good to go

File: tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py (hunk 5 above)
Comment (BlessedDev) [reply tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py:45]: Thanks for catching that. It was a mistake and I have corrected the order.

**Review by leohemsted — APPROVED** [review]

**Pull request merged**